File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        if heights == 2:
            return 1 * min(heights[0], heights[1])
        left = 0
        right  = len(heights) - 1
        maxSum = 0
        while left < right:
            #print(f"Left Num: {heights[left]}")
            #print(f"Right Num: {heights[right]}")
            #print(f"New Sum: {(right - left) * min(heights[left], heights[right])}")
            if ((right - left) * min(heights[left], heights[right])) > maxSum:
                maxSum = ((right - left) * min(heights[left], heights[right]))
                #print(f"New Max Sum: {maxSum}")
            if heights[left] <= heights[right]:
                left += 1
            elif heights[left] >= heights[right]:
                right -= 1
        return maxSum

        

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.cpp
#include <iostream>
#include <unordered_map>
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        std::unordered_map <char, int> charTrack;
        if (s.length() <= 1) {
            return s.length();
        }
        int left = 0;
        int right = 0;
        int maxCount = 0;
        int curCount = 0;
        while (right < s.length()) {
            if (charTrack.contains(s[right]) && charTrack[s[right]] == 1) {
                maxCount = max(maxCount, curCount);
                charTrack.erase(s[left]);
                left += 1;
                curCount -= 1;
                continue;
            }
            charTrack[s[right]] = 1;
            curCount += 1;
            right += 1;
        }
        return max(maxCount, curCount);
    }
};
